spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -529,114 +400,131 @@ function onRecvPacket(packet: Packet) {
+    if len(data.forwarding.hops) == 1 {
       // we're on the last hop, we can set memo and clear
       // the next forwarding
-      memo = forwarding.memo
+      memo = data.forwarding.memo
       nextForwarding = nil
     }
     // send the tokens we received above to the next port and channel
     // on the forwarding path
     // and reduce the forwarding by the first element
-    packetSequence = sendFungibleTokens(
-      receivedTokens,
-      receiver, // sender of next packet
-      finalReceiver, // receiver of next packet
-      memo,
-      nextForwarding,
-      forwarding.hops[0].portId,
+  
+  // Here we must call the core sendPacket providing the correct forwardingPayload --> Need to construct the payload 
+  //construct payload 
+
+  forwardingPayload= FungibleTokenPacketDataV2 {
+    tokens: receivedTokens,
+    sender: receiver
+    receiver: finalReceiver
+    memo: memo, 
+    // a struct containing the list of next hops, 
+    // determining where the tokens must be forwarded next, 
+    // and the memo for the final hop
+    forwarding: nextForwarding 
+  }
+  
+  packetSequence=handler.sendPacket(
       forwarding.hops[0].channelId,
-      Height{},
       currentTime() + DefaultHopTimeoutPeriod,
+      forwardingPayload
     )
     // store packet for future sending ack
-    privateStore.set(packetForwardPath(forwarding.hops[0].portId, forwarding.hops[0].channelId, packetSequence), packet)
+    privateStore.set(packetForwardPath(forwarding.hops[0].channelId, packetSequence), sequence)
     // use async ack until we get successful acknowledgement from further down the line.
-    return nil
+    return nil, true
   }
 
-  return ack
+  return ack,true
 }
 ```
 
 `onAcknowledgePacket` is called by the routing module when a packet sent by this module has been acknowledged.
 
 ```typescript
 function onAcknowledgePacket(
-  packet: Packet,
-  acknowledgement: bytes) {
+  sourceChannelId: bytes,
+  destChannelId: bytes, 
+  sequence: bigEndianUint64,
+  payload: Payload, 
+  acknowledgement: bytes
+  ): bool {
   // if the transfer failed, refund the tokens
   // to the sender account. In case of a packet sent for a
   // forwarded packet, the sender is the forwarding
   // address for the destination channel of the forwarded packet.
   if !(acknowledgement.success) {
-    refundTokens(packet)
+    refundTokens(sourceChannelId, payload)
   }
 
   // check if the packet that was sent is from a previously forwarded packet
-  prevPacket = privateStore.get(packetForwardPath(packet.sourcePort, packet.sourceChannel, packet.sequence))
+  prevPacketSeq = privateStore.get(packetForwardPath(sourceChannelId, sequence))
 
-  if prevPacket != nil {
+  if prevPacketSeq != nil {
     if acknowledgement.success {
       FungibleTokenPacketAcknowledgement ack = FungibleTokenPacketAcknowledgement{true, "forwarded packet succeeded"}
       handler.writeAcknowledgement(
-        prevPacket,
+        destChannelId,
```

Comment:
We aren't using the destChannelId here. We need the dest channel of the **previous packet**

If its not available here then we need to store it along with the previous sequence.